remove database backend #489

joshdimanteto opened the pull request:
This PR will close #489

## Description
Remove the database backend functionality from the DataGateway API. It includes significant changes such as deleting database-related files and functionality, updating tests to remove database dependencies, and modifying the configuration to align with the changes.

## Testing Instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check GitHub Actions build
- [ ] If `icatdb Generator Script Consistency Test` CI job fails, is this because of a deliberate change made to the script to change generated data (which isn't actually a problem) or is here an underlying issue with the changes made?
- [ ] Review changes to test coverage
- [ ] Does this change mean a new patch, minor or major version should be made? If so, does one of the commit messages feature `fix:`, `feat:` or `BREAKING CHANGE:` so a release is automatically made via GitHub Actions upon merge?
- [ ] {more steps here}

## Agile Board Tracking
Connect to #489


---
## notes

The icat models still use SQLAlchemy, this is going to be update in the pydantic V2 upgrade to use pydantic

joshdimanteto committed:
refactor: remove database backend #489

joshdimanteto committed:
fix intergration tests #489

joshdimanteto committed:
set up icat stack before pip install test #489

joshdimanteto committed:
ignore vulnerability for werkzeug and cryptography #489

joshdimanteto committed:
remove remaining remnants of the backends #489

joshdimanteto committed:
update python to use Python 3.9.12 for Ubuntu 24.04 compatibility

joshdimanteto committed:
Update lock file

joshdimanteto committed:
update lock file

joshdimanteto committed:
pin python version to 3.10

joshdimanteto committed:
mend

joshdimanteto committed:
upgrade the black verison

joshdimanteto committed:
reformat code using latest version black #517

joshdimanteto committed:
fix linting errors #517

joshdimanteto committed:
upgrade faker #517

- remove downgrade of setuptools

joshdimanteto committed:
upgrade the version of poetry to 1.8.0 #517

joshdimanteto committed:
remove install with constraints #517

joshdimanteto committed:
revert faker upgrade #517

joshdimanteto committed:
revert icat_db_generator changes #517

joshdimanteto committed:
run black formatter for icat_db_generator.py #517

joshdimanteto committed:
fix linting in icat_db_generator.py #517

joshdimanteto committed:
ignore B907 in multiple SQL query constructions

Added `# noqa: B907` in several places where manual quoting is required
for valid SQL query syntax. Using `!r` would change the format and break
query parsing, so explicit quotes are necessary. This ensures tests and
API calls remain correct while satisfying linting requirements.

joshdimanteto committed:
change strict=True to False for zip functions #517

joshdimanteto committed:
Change fixture for db to "function" from "package"

- In the test_query_filter_factory the test don't work on the CI as it
  is not setting up the db backend.

joshdimanteto committed:
reset ICAT client before each test to avoid shared state issues #517

joshdimanteto committed:
silence flake8 B909 warnings in query filter factory test #517

joshdimanteto committed:
Merge branch 'remove-older-python-versions-#517' into remove-database-backend-#489

joshdimanteto committed:
Merge branch 'remove-older-python-versions-#517' into remove-database-backend-#489

joshdimanteto committed:
change strict flag to true in helpers #517

joshdimanteto committed:
Merge branch 'remove-older-python-versions-#517' into remove-database-backend-#489

joshdimanteto commented on .github/workflows/ci-build.yml:
```diff
@@ -353,13 +353,67 @@ jobs:
+          mv -f /home/runner/install/icat.server/run.properties.tmp /home/runner/install/icat.server/run.properties
+      - name: Reinstall ICAT Server
+        run: |
+          cd /home/runner/install/icat.server/ && ./setup -vv install
```
check if this is still needed

joshdimanteto replied:
it is, as it needs an ICAT server

VKTB replied:
Just wondering why it was working before without any of this but now it needs all the ICAT stuff?

joshdimanteto replied:
This work before as it was using the db backend instead of the python icat backend so it didn't need the extra setup

VKTB replied:
That makes sense, thanks.

joshdimanteto committed:
remove unused sql imports #489

joshdimanteto commented on test/integration/conftest.py:
```diff
@@ -65,11 +61,11 @@ def valid_db_credentials_header():
     yield {"Authorization": f"Bearer {session.id}"}
 
-    delete_row_by_id(SESSION, "Test")
+    # delete_row_by_id(SESSION, "Test")
```
```suggestion

yield {"Authorization": f"Bearer {session.id}"}

```

joshdimanteto commented on test/integration/conftest.py:
```diff
@@ -45,11 +42,10 @@ def flask_test_app():
 def flask_test_app_db():
```
This name should change. I have changed in #524 where i had to refactored the tests

VKTB commented on datagateway_api/src/common/config.py:
Once migrated to pydantic v2, I would recommend that [pydantic-settings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/) is used for managing the config.

joshdimanteto replied:
This work before as it was using the db backend instead of the python icat backend so it didn't need the extra setup

joshdimanteto replied:
I will create a separated pr to do this ([pydantic-settings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/)[pydantic-settings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/)). I will do this pr after the port to fast api pr #524

joshdimanteto replied:
#526

VKTB commented on datagateway_api/src/common/config.py:
```diff
@@ -93,44 +72,25 @@ def require_db_config_value(cls, value, values):  # noqa: B902, N805
-            raise TypeError("field required")
+        if value is None:
+            raise TypeError("Field required for `python_icat`.")
         return value
```
Shouldn't the fields in the `@validator` signature now always be mandatory? If so I don't the point of having a validator, just the definition of the fields need updating.

joshdimanteto replied:
I've done that in the next pr but i'll copy the change into this pr https://github.com/ral-facilities/datagateway-api/pull/521/changes#diff-22481ddb482cc5158876a842b78be896b32d69bac61530202b4a831e08a5c442

joshdimanteto committed:
address review comments #489

VKTB reviewed (approved):
I think the changes look fine, I tried my best. I have not tested it locally as I don't have a working stack to point it it to so I just trust that it works as expected. Might be worth asking someone else to have a look too. Approving.

patrick-austin commented on .github/workflows/ci-build.yml:
```diff
@@ -350,16 +350,72 @@ jobs:
+        uses: actions/setup-java@b36c23c0d998641eff861008f374ee103c25ac73 # v4.4.0
+        with:
+          distribution: "temurin"
+          java-version: 8
```
Given the CI for icat.server runs on 11/21 and we're looking to run 25 in production for Diamond, we can probably increase this (not that it should really change anything).

patrick-austin replied:
Alan says 11 would be best.

patrick-austin replied:
He also says that actually this won't matter as icat-ansible will install java itself, so we don't need this step at all (big if true).

patrick-austin replied:
He also says to use the payara7 branch

patrick-austin commented on .github/workflows/ci-build.yml:
```diff
@@ -350,16 +350,72 @@ jobs:
+
+        # ICAT Ansible clone and install dependencies
+      - name: Checkout icat-ansible
+        if: success()
```
```suggestion
```

As far as I know this doesn't do anything, since the [default behaviour is to only run the step if the previous ones succeeded](https://docs.github.com/en/actions/reference/workflows-and-actions/expressions#status-check-functions) - unless I'm missing something here?

patrick-austin commented on datagateway_api/src/common/config.py:
```diff
@@ -48,89 +48,28 @@ class UseReaderForPerformance(BaseModel):
 
         # Enables assignment validation on the BaseModel fields. Useful for when the
-        # backend type is changed using the set_backend_type function.
         validate_assignment = True
```
Is setting `validate_assignment = True` still necessary at all? `set_backend_type` is gone, and searching the repo I can't see any instances where we assign a new variable to a config field, so unless there's something subtle going on, I think this can be removed?

At the very least, can edit the comment so that it's a complete sentence and doesn't end [`when the`](https://www.reddit.com/r/whenthe/)

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -125,10 +129,7 @@ def test_invalid_query_creation(self, icat_client):
-                    "title": "like '%Test data for the Python ICAT Backend on"
-                    " DataGateway API%'",
-                },
+                {"title": "like '%Test data for Python ICAT on" " DataGateway API%'"},
```
```suggestion
                {"title": "like '%Test data for Python ICAT on DataGateway API%'"},
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -144,17 +145,13 @@ def test_invalid_query_creation(self, icat_client):
                         "summary": None,
-                        "title": "Test data for the Python ICAT Backend on DataGateway"
-                        " API 0",
+                        "title": "Test data for Python ICAT on DataGateway" " API 0",
```
```suggestion
                        "title": "Test data for Python ICAT on DataGateway API 0",
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -144,17 +145,13 @@ def test_invalid_query_creation(self, icat_client):
-                    "title": "like '%Test data for the Python ICAT Backend on"
-                    " DataGateway API%'",
-                },
+                {"title": "like '%Test data for Python ICAT on" " DataGateway API%'"},
```
```suggestion
                {"title": "like '%Test data for Python ICAT on DataGateway API%'"},
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -168,8 +165,7 @@ def test_invalid_query_creation(self, icat_client):
                         "summary": None,
-                        "title": "Test data for the Python ICAT Backend on DataGateway"
-                        " API 0",
+                        "title": "Test data for Python ICAT on DataGateway" " API 0",
```
```suggestion
                        "title": "Test data for Python ICAT on DataGateway API 0",
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -185,10 +181,7 @@ def test_invalid_query_creation(self, icat_client):
-                    "title": "like '%Test data for the Python ICAT Backend on"
-                    " DataGateway API%'",
-                },
+                {"title": "like '%Test data for Python ICAT on" " DataGateway API%'"},
```
```suggestion
                {"title": "like '%Test data for Python ICAT on DataGateway API%'"},
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -198,10 +191,7 @@ def test_invalid_query_creation(self, icat_client):
-                    "title": "like '%Test data for the Python ICAT Backend on"
-                    " DataGateway API%'",
-                },
+                {"title": "like '%Test data for Python ICAT on" " DataGateway API%'"},
```
```suggestion
                {"title": "like '%Test data for Python ICAT on DataGateway API%'"},
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -217,54 +207,38 @@ def test_invalid_query_creation(self, icat_client):
                         "summary": None,
-                        "title": "Test data for the Python ICAT Backend on DataGateway"
-                        " API 0",
+                        "title": "Test data for Python ICAT on DataGateway" " API 0",
```
```suggestion
                        "title": "Test data for Python ICAT on DataGateway API 0",
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -217,54 +207,38 @@ def test_invalid_query_creation(self, icat_client):
-                    "title": "like '%Test data for the Python ICAT Backend on"
-                    " DataGateway API%'",
-                },
+                {"title": "like '%Test data for Python ICAT on" " DataGateway API%'"},
```
```suggestion
                {"title": "like '%Test data for Python ICAT on DataGateway API%'"},
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -217,54 +207,38 @@ def test_invalid_query_creation(self, icat_client):
-                        " API 0",
-                    },
-                ],
+                [{"title": "Test data for Python ICAT on DataGateway" " API 0"}],
```
```suggestion
                [{"title": "Test data for Python ICAT on DataGateway API 0"}],
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -217,54 +207,38 @@ def test_invalid_query_creation(self, icat_client):
-                    "title": "like '%Test data for the Python ICAT Backend on"
-                    " DataGateway API%'",
-                },
+                {"title": "like '%Test data for Python ICAT on" " DataGateway API%'"},
```
```suggestion
                {"title": "like '%Test data for Python ICAT on DataGateway API%'"},
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -217,54 +207,38 @@ def test_invalid_query_creation(self, icat_client):
                     {
-                        "title": "Test data for the Python ICAT Backend on DataGateway"
-                        " API 0",
+                        "title": "Test data for Python ICAT on DataGateway" " API 0",
```
```suggestion
                        "title": "Test data for Python ICAT on DataGateway API 0",
```

patrick-austin commented on test/integration/datagateway_api/icat/test_query.py:
```diff
@@ -217,54 +207,38 @@ def test_invalid_query_creation(self, icat_client):
-                    "title": "like '%Test data for the Python ICAT Backend on"
-                    " DataGateway API%'",
-                },
+                {"title": "like '%Test data for Python ICAT on" " DataGateway API%'"},
```
```suggestion
                {"title": "like '%Test data for Python ICAT on DataGateway API%'"},
```

patrick-austin commented on README.md:
```diff
@@ -359,10 +343,9 @@ require an ICAT stack. In order to cover all the code you will need to run both
-tests (containing tests for backend specific components, including tests for the
-different types of endpoints) and Database Backend tests (like the ICAT backend tests,
-but covering only the most used aspects of the API).
+The tests are split into 2 main sections: non Python ICAT specific (testing features such as the date handler) and python ICAT
```
```suggestion
The tests are split into 2 main sections: non Python ICAT specific (testing features such as the date handler) and Python ICAT
```

patrick-austin commented on README.md:
```diff
@@ -438,9 +421,8 @@ The logic for each endpoint is within `/src/resources` - they're split into enti
-Backend uses this for its mapped classes (explained below) and the Python ICAT Backend
-uses this for interacting with ICAT objects within Python ICAT. In most instances, the
+maps entity names to their field name.
+this is used for interacting with ICAT objects within Python ICAT. In most instances, the
```
```suggestion
This is used for interacting with ICAT objects within Python ICAT. In most instances, the
```

patrick-austin commented on README.md:
```diff
@@ -464,16 +446,13 @@ handler can be used to convert dates between string and datetime objects (using
 in a request body is a date, at which point it can be converted to a datetime object,
-ready for storing in ICAT. The handler is currently only used in the Python ICAT
-Backend, however this is non-backend specific class.
+ready for storing in ICAT. The handler is currently only used in the Python ICAT.
```
```suggestion
ready for storing in ICAT.
```

patrick-austin commented on README.md:
```diff
@@ -794,12 +725,8 @@ containing over 300 requests, with each type of endpoint for every entity as wel
-for the Database Backend, as logging in using that backend is slightly different to
-logging in via the Python ICAT Backend.
+This collection is mainly based around the Python ICAT (request bodies for
+creating and updating data uses camelCase attribute names as accepted by that python_icat).
```
For me it doesn't really make sense to say it's mainly based around Python ICAT, at is it implies it is partially based around something else (i.e. the DB) that no longer exists at all - but I don't think it's a big deal either way.

joshdimanteto replied:
I am going to removed those two line as datagateway-api only use python icat now

patrick-austin reviewed (changes requested):
Functionally I think it's OK (I can run the API and get results back).

There's a few whitespace changes I would make (as a result of strings being moved onto one line) and one or two possible code changes.

joshdimanteto committed:
address review comments #489

- remove unnecessary argument "validate_assignment = True" in config
- format strings better in test_query for title
- update README

joshdimanteto committed:
use payara7 branch for icat-ansible #489

- remove installation of java this should installed by the ansible
  script

joshdimanteto committed:
remove unnecessary "success()" for ci_build #489

patrick-austin reviewed (approved):
As discussed over Teams, there are some instances where the models do not strictly represent the types/fields required by the API (for example required fields when POSTing or id when PATCHing).

However, given that further changes are in progress with the FastAPI refactor and the primary use case for the API is DataGateway's read operations, I'm happy that the use of Optional lets things work for now, and it doesn't matter so much that someone referring to the Swagger wouldn't know id is required for PATCH (as who's looking at the Swagger in the first place?)

joshdimanteto committed:
Merge branch 'develop' into remove-database-backend-#489

joshdimanteto committed:
Merge branch 'develop' into remove-database-backend-#489
